fix(logging): queue_listener handler for Python >= 3.12

jderrien opened the pull request:
<!--
By submitting this pull request, you agree to:
- follow [Litestar's Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)
- follow [Litestar's contribution guidelines](https://github.com/litestar-org/.github/blob/main/CONTRIBUTING.md)
- follow the [PSFs's Code of Conduct](https://www.python.org/psf/conduct/)
-->
## Description

- Fix the `queue_listener` handler for Python 3.12

Python 3.12 introduced a new way to configure `QueueHandler` and `QueueListener` via `logging.config.dictConfig()`. This described [here](https://docs.python.org/3/library/logging.config.html#configuring-queuehandler-and-queuelistener).

The listener still needs to be started & stopped, as previously. To do so, I've introduced `LoggingQueueListener`.

And as stated in the doc:

> Any custom queue handler and listener classes will need to be defined with the same initialization signatures as [QueueHandler](https://docs.python.org/3/library/logging.handlers.html#logging.handlers.QueueHandler) and [QueueListener](https://docs.python.org/3/library/logging.handlers.html#logging.handlers.QueueListener).

<!--
Please add in issue numbers this pull request will close, if applicable
Examples: Fixes #4321 or Closes #1234

Ensure you are using a supported keyword to properly link an issue:
https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
## Closes

Closes #2954

provinzkraut reviewed (approved):
@JacobCoffee Can you take a look at this? Looks good from my side.

provinzkraut commented:
@jderrien Thanks for the PR :)
Can you expand the PR description a bit, e.g. how this changes fixes the referenced issue?

jderrien commented:
> @jderrien Thanks for the PR :) Can you expand the PR description a bit, e.g. how this changes fixes the referenced issue?

I've updated the description, don't hesitate to comment the PR if you need more explainations. :)

Also, please wait before merging this. It seems some test jobs hung on on GitHub Actions. It runs well locally since I added `cleanup_logging()`, I'm not sure what's going on. :/

jderrien commented:
> It seems some test jobs hung on on GitHub Actions

https://github.com/litestar-org/litestar/actions/runs/8238535303
https://github.com/litestar-org/litestar/actions/runs/8250579528

The test suite on Python 3.12 takes ~10 min and gets killed.

While I didn't have this issue on macOS, I'll try to run the test suite in a Linux container tomorrow. Hopefully I'll be able to reproduce the issue.

peterschutt commented:
I did some testing yesterday and found it to be flaky, sometimes it would hang and sometimes not. I re-requested the 3.12 tests in CI and they passed that time which confirms it. Unfortunately, when it did hang I'd get different behaviors after hitting `^C`, sometimes the tests would continue and pass, sometimes I'd get errors out of pytest plugins, and sometimes errors out of low level threading stuff, but nothing that would actually point to any of our code :/

jderrien commented on litestar/logging/config.py:
```diff
@@ -49,7 +49,17 @@
+    default_handlers["queue_listener"].update(
+        {
+            "class": "logging.handlers.QueueHandler",
+            "queue": Queue(-1),
```
Instantiating a `queue.Queue` object in the handler config might be the source of the "test suite hanging" problem.

peterschutt replied:
Perhaps - maybe try marking all of those logging tests with `pytest.mark.xdist("something")` so they all run in the same process and see if that helps?

jderrien replied:
In fact, I guess `litestar/logging/config.py` and so `default_handlers` might be loaded in different tests and context. Instanciating a `Queue` object directly in the config seems to be a bad idea in most cases.

Instead, as mentioned in the Python doc, I can also use a factory. From what I've seen locally, it fixes the problem. I'm going to push the change, but not sure how I can run the test suite multiple times on Github Actions to check if everything is fine... I can just amend the commit without any change to refresh the commit id. But maybe you'd have a better idea?

provinzkraut replied:
> I can just amend the commit without any change to refresh the commit id. But maybe you'd have a better idea?

You can also just push an empty commit if you want to trigger a rerun

peterschutt reviewed (changes requested):
Flaky test.

Just adding this so we don't accidentally merge.

jderrien commented:
> Flaky test.
> 
> Just adding this so we don't accidentally merge.

@guacs advised me to use Github Actions directly on my fork, so I did.

I worked on [a specific branch](https://github.com/jderrien/litestar/tree/fix-logging-queue-listener-handler-testing), the only change is [this commit](https://github.com/jderrien/litestar/commit/c1cee6521728cd57d1707640ffcf8a0d8b73b32f).

On [these 25 runs](https://github.com/jderrien/litestar/actions/workflows/ci.yml) (26 but the first one failed due to a config mistake), everything looks good, at least on Python 3.12.

There were 3 failures on Python 3.11 on `tests/unit/test_channels/test_plugin.py::test_create_ws_route_handlers_arbitrary_channels_allowed` (which is already marked as flaky).

The problem seems to be solved.

jderrien commented on litestar/logging/standard.py:
```diff
@@ -18,7 +18,7 @@ class QueueListenerHandler(QueueHandler):
 
         def __init__(self, handlers: list[Any] | None = None) -> None:
-            """Initialize `?QueueListenerHandler`.
+            """Initialize `QueueListenerHandler`.
```
Was this a typo?

[The doc preview](https://litestar-org.github.io/litestar-docs-preview/3185/reference/logging/standard.html) seems to be broken, compared to the [official one](https://docs.litestar.dev/2/reference/logging/standard.html).

JacobCoffee replied:
Im looking into this

JacobCoffee replied:
@provinzkraut im not sure what makes it not load inside the preview, or if on next docs deploy the production docs would also stop loading?

I cant get them to load in any sort of configuration of Python version venvs and other means...
https://paste.pythondiscord.com/FKRQ is what I have for adding some notes to this, but i cant get our custom `QueueListenerHandler` to document itself in any way. Maybe it will be fine

provinzkraut replied:
It's because this is hidden in an if block 🤷

jderrien replied:
@provinzkraut the `if` block is already on the `main` branch and the doc is correct? 🤔

provinzkraut replied:
But the docs build on `3.12`. For `3.12`, the `QueueListenerHandler` isn't defined. This is a general issue and a breaking change by the way. The previous version had `QueueListenerHandler` defined in both branches.

jderrien replied:
@provinzkraut I guess I understand better.

But it means [the official doc](https://docs.litestar.dev/latest/reference/logging/standard.html) has been generated with Python < 3.12 since it refers to the former `QueueListenerHandler` class (not to the aliased one), right?

Or does it mean the official doc is not up-to-date (not generated/published recently)?

While I understand this is a breaking change at the API level, I'm not sure what I should do next. Any advice?

provinzkraut replied:
I'm not 100% sure, but I think the issue is that the `LoggingQueueListener` isn't in `__all__` and what *is* in `__all__` (`QueueListenerHandler`), isn't available on < 3.12. My guess would be that if you rename ``LoggingQueueListener` > `QueueListenerHandler`, it should work. If not, then I'm wrong and the issue is something else

jderrien replied:
You're right about `__all__`.

That's said, renaiming `LoggingQueueListener` to `QueueListenerHandler` is not a good solution IMO, because this is a `QueueListener` object and not a `logging.Handler` one. I think naming is important. Also, the constructor args are not the same and I don't think I can change that... So it would be confusing, and the API doc would be incorrect.

I could keep `QueueListenerHandler = QueueHandler`, but it won't help to get `QueueListenerHandler` fully documented (for Python < 3.12).

I could remove the `if` condition, it shouldn't be a problem (technically). Even if it cannot be used with Python >= 3.12, it can be documented. Or I can keep the `if` condition and redefine `QueueListenerHandler` with the same constructor signature, and raise an exception (or a warning) stating it cannot be used with Python > 3.12.

I'm not sure how to manage this at the project level.

I tried to take a second look to check if we could keep `QueueListenerHandler` around. 👀 And from what I've understood, the signature has to be changed so it would break the API anyway. This is because the handler is instanciated with a single argument, which is the queue (and not `handlers: list[Any] | None`): https://github.com/python/cpython/blob/v3.12.2/Lib/logging/config.py#L744 -- `_configure_queue_handler` has been introduced in Python 3.12.

provinzkraut replied:
I think for now we'll have to just keep the `QueueListenerHandler = QueueHandler` around somehow. `QueueListenerHandler` *and* `LoggingQueueListener` need to always be defined, no matter what Python version is being used. They should also be both in `_all__`. 

> I could remove the if condition, it shouldn't be a problem (technically). Even if it cannot be used with Python >= 3.12, it can be documented.

I think that's the best approach for now.

jderrien replied:
Sorry if all of this is confusing, but removing the `if` statement AND keeping `QueueListenerHandler = QueueHandler` won't work. :)

To clarify (and if I'm not mistaken):

Currently (`main` branch):
⛔️ The default `queue_listener` handler is broken on Python 3.12
⛔️ Generating the API doc shows `QueueListenerHandler` as an alias of `logging.QueueHandler`, which is not the case for Python < 3.12
⚠️ Official API doc still shows the previous implementation (Python < 3.12) for an unknown reason (?)
⛔️ When upgrading to Python 3.12, the constructor signature of`QueueListenerHandler` is not the same anymore, which implies a breaking change for any "custom" usage of `QueueListenerHandler`

Keeping `QueueListenerHandler = QueueHandler` won't help to get a correct API doc for Python < 3.12.

I guess removing the `if` statement seems better:
✅ The default `queue_listener` handler will work out of the box
✅ API doc will show the informations for `QueueListenerHandler` (and `LoggingQueueListener`)
✅ The fact it doesn't work with Python >= 3.12 can be documented
✅ `QueueListenerHandler` will be able to be deprecated later (maybe we could/should add a warning right now?)
⛔️ Any usage relying on `QueueListenerHandler = QueueHandler` will break (that's said, as of today, the official API doc doesn't mention this aliasing on Python >= 3.12) -- I guess it might be a problem mostly if some custom configs have been put in place because it was not working out of the box with Python 3.12.
⛔️ Usage of `QueueListenerHandler` will break when upgrading to Python 3.12 for any custom implementation using it. But it can be documented. And using `QueueListenerHandler = QueueHandler` wouldn't help, because the constructor signatures are not the same (as stated before).

In all cases, there is a breaking change we can't avoid due to Python 3.12. So the best we can do is making it clear. Also, I'm not sure anyone is using `QueueListenerHandler` outside of the `default_handlers` provided, so the change might be minimal (that's said, I'm new to Litestar).

jderrien replied:
@provinzkraut I updated the PR, preview doc is available [here](https://litestar-org.github.io/litestar-docs-preview/3185/reference/logging/standard.html).

Any recommendation?

provinzkraut replied:
Thanks for the great summary @jderrien! The current implementation seems like the best solution to me. Let's just keep it like that.

JacobCoffee commented on litestar/logging/standard.py:
```diff
@@ -31,4 +31,18 @@ def __init__(self, handlers: list[Any] | None = None) -> None:
-    QueueListenerHandler = QueueHandler
+
+    class LoggingQueueListener(QueueListener):
+        """Custom `QueueListener` which starts and stops the listening process."""
```
```suggestion
        """Custom ``QueueListener`` which starts and stops the listening process."""
```

JacobCoffee commented on litestar/logging/standard.py:
```diff
@@ -31,4 +31,18 @@ def __init__(self, handlers: list[Any] | None = None) -> None:
+        """Custom `QueueListener` which starts and stops the listening process."""
+
+        def __init__(self, queue: Queue[LogRecord], *handlers: Handler, respect_handler_level: bool = False) -> None:
+            """Initialize `LoggingQueueListener`.
```
```suggestion
            """Initialize ``LoggingQueueListener``.
```

JacobCoffee commented on litestar/logging/standard.py:
```diff
@@ -31,4 +31,18 @@ def __init__(self, handlers: list[Any] | None = None) -> None:
+            Args:
+                queue: The queue to send messages to
+                *handlers: A list of handlers which will handle entries placed on the queue
+                respect_handler_level: If respect_handler_level is True, a handler's level is respected (compared with the level for the message) when deciding whether to pass messages to that handler
```
```suggestion
                respect_handler_level: If ``respect_handler_level`` is ``True``, a handler's level is respected (compared with the level for the message) when deciding whether to pass messages to that handler
```

JacobCoffee reviewed (approved):
lgtm aside from [weird page load](https://litestar-org.github.io/litestar-docs-preview/3185/reference/logging/standard.html)

peterschutt commented:
> > Flaky test.
> > Just adding this so we don't accidentally merge.
> 
> @guacs advised me to use Github Actions directly on my fork, so I did.
> 
> I worked on [a specific branch](https://github.com/jderrien/litestar/tree/fix-logging-queue-listener-handler-testing), the only change is [this commit](https://github.com/jderrien/litestar/commit/c1cee6521728cd57d1707640ffcf8a0d8b73b32f).
> 
> On [these 25 runs](https://github.com/jderrien/litestar/actions/workflows/ci.yml) (26 but the first one failed due to a config mistake), everything looks good, at least on Python 3.12.
> 
> There were 3 failures on Python 3.11 on `tests/unit/test_channels/test_plugin.py::test_create_ws_route_handlers_arbitrary_channels_allowed` (which is already marked as flaky).
> 
> The problem seems to be solved.

Fantastic! Thanks very much!

jderrien commented on litestar/logging/standard.py:
```diff
@@ -1,34 +1,45 @@
+        """
+        super().__init__(Queue(-1))
+        handlers = resolve_handlers(handlers) if handlers else [StreamHandler()]
+        self.listener = LoggingQueueListener(self.queue, *handlers)  # type: ignore[arg-type]
```
mypy was returning:
```
litestar/logging/standard.py:45: error: Argument 1 to "LoggingQueueListener" has incompatible type "_QueueLike[Any]"; expected "Queue[LogRecord]"  [arg-type]
```

I don't know if I could do something better here. It seems I cannot import `_QueueLike`.

provinzkraut replied:
I think this is fine. It seems that `QueueHandler.queue` isn't typed properly as `Queue[LogRecord]` internally, so nothing we can fix here.

jderrien replied:
@provinzkraut I guess I could type `LoggingQueueListener.queue` as `Any` and remove `type: ignore[arg-type]` here. But not sure it would be better.

provinzkraut replied:
I think that'd be more confusing actually

provinzkraut commented:
@JacobCoffee Good to go from your end?
